src/distributors/base-installer.ts

```diff
@@ -0,0 +1,121 @@
+import * as tc from '@actions/tool-cache';
+import * as core from '@actions/core';
+import semver from 'semver';
+import path from 'path';
+import * as httpm from '@actions/http-client';
+import { getVersionFromToolcachePath } from '../util';
+
+export interface JavaInstallerOptions {
+    version: string;
+    arch: string;
+    javaPackage: string;
+}
+
+export interface JavaInstallerResults {
+    javaVersion: string;
+    javaPath: string;
+}
+
+export interface IJavaRelease {
+    resolvedVersion: string;
+    link: string;
+}
+
+export abstract class JavaBase {
+    protected http: httpm.HttpClient;
+    protected version: string;
+    protected arch: string;
+    protected javaPackage: string;
+    constructor(protected distributor: string, initOptions: JavaInstallerOptions) {
+        this.http = new httpm.HttpClient('setup-java', undefined, {
+            allowRetries: true,
+            maxRetries: 3
+          });
+          this.version = this.normalizeVersion(initOptions.version);
+          this.arch = initOptions.arch;
+          this.javaPackage  = initOptions.javaPackage;
+    }
+
+    protected abstract downloadTool(javaRelease: IJavaRelease): Promise<JavaInstallerResults>;
+    protected abstract resolveVersion(range: semver.Range): Promise<IJavaRelease>;
+
+    public async getJava(): Promise<JavaInstallerResults> {
+        const range = new semver.Range(this.version);
+        let foundJava = this.findInToolcache(range);
+
+        if(!foundJava) {
+            // download Java if it is not found locally
+            const javaRelease = await this.resolveVersion(range)
+            foundJava = await this.downloadTool(javaRelease);
+        }
+
+        this.setJavaDefault(foundJava.javaPath);
+
+        return foundJava;
+    }
+
+    protected get toolcacheFolderName(): string {
+        return `Java_${this.distributor}_${this.javaPackage}`;
+    }
+
+    protected findInToolcache(version: semver.Range): JavaInstallerResults | null {
+        const toolPath = tc.find(this.toolcacheFolderName, version.raw, this.arch);
+        if (!toolPath) {
+            return null;
+        }
+
+        return {
+            javaVersion: getVersionFromToolcachePath(toolPath),
+            javaPath: toolPath
+        };
+    }
+
+    protected setJavaDefault(toolPath: string) {
+        core.exportVariable('JAVA_HOME', toolPath);
+        core.addPath(path.join(toolPath, 'bin'));
+        core.setOutput('path', toolPath);
+        core.setOutput('version', this.version);
```

Comment:
I think we should set resolved version here from `foundJava`

Reply:
Done.